File: wear/src/main/java/com/example/sobti/MainActivity.java
package com.example.sobti;

import android.Manifest;
import android.content.pm.PackageManager;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.FragmentActivity;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.wearable.*;
import java.util.Random;

public class MainActivity extends FragmentActivity implements SensorEventListener {

    private static final int PERMISSION_REQUEST_CODE = 100;
    private static final String DATA_PATH = "/health_data";

    private TextView tvHeartRate, tvSteps, tvStatus;
    private SensorManager sensorManager;
    private Sensor heartRateSensor;
    private Sensor stepCounterSensor;

    private int currentHeartRate = 0;
    private int currentSteps = 0;

    private DataClient dataClient;

    // Test/Fake heart rate simulation
    private boolean useTestMode = true; // Set to false for real sensor data
    private Random random = new Random();
    private int testHeartRate = 75;
    private int testTrendCounter = 0;
    private boolean testIncreasing = false;
    private boolean testDecreasing = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();
        requestPermissions();
        setupSensors();

        dataClient = Wearable.getDataClient(this);

        // Start test mode if enabled
        if (useTestMode) {
            startTestMode();
        }
    }

    private void initViews() {
        tvHeartRate = findViewById(R.id.tvHeartRate);
        tvSteps = findViewById(R.id.tvSteps);
        tvStatus = findViewById(R.id.tvStatus);
    }

    private void requestPermissions() {
        String[] permissions = {
            Manifest.permission.BODY_SENSORS,
            Manifest.permission.ACTIVITY_RECOGNITION
        };

        ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_CODE);
    }

    private void setupSensors() {
        sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);

        if (sensorManager != null) {
            heartRateSensor = sensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE);
            stepCounterSensor = sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER);

            if (heartRateSensor != null && !useTestMode) {
                sensorManager.registerListener(this, heartRateSensor, SensorManager.SENSOR_DELAY_NORMAL);
                tvStatus.setText("Heart Rate: Active");
            } else {
                tvStatus.setText("Test Mode: Active");
            }

            if (stepCounterSensor != null) {
                sensorManager.registerListener(this, stepCounterSensor, SensorManager.SENSOR_DELAY_NORMAL);
            }
        }
    }

    private void startTestMode() {
        // Simulate heart rate readings every 3 seconds
        new Thread(() -> {
        while (useTestMode) {
            try {
                Thread.sleep(3000);

                // Randomly decide to simulate abnormal patterns
                if (random.nextInt(20) == 0 && !testIncreasing && !testDecreasing) {
                    // Start increasing trend
                    testIncreasing = true;
                    testTrendCounter = 0;
                } else if (random.nextInt(20) == 1 && !testIncreasing && !testDecreasing) {
                    // Start decreasing trend
                    testDecreasing = true;
                    testTrendCounter = 0;
                }

                // Generate heart rate based on current trend
                if (testIncreasing) {
                    testHeartRate += random.nextInt(8) + 5; // Increase by 5-12
                    testTrendCounter++;

                    if (testTrendCounter >= 5 || testHeartRate > 140) {
                        testIncreasing = false;
                        testTrendCounter = 0;
                    }
                } else if (testDecreasing) {
                    testHeartRate -= random.nextInt(8) + 5; // Decrease by 5-12
                    testTrendCounter++;

                    if (testTrendCounter >= 5 || testHeartRate < 40) {
                        testDecreasing = false;
                        testTrendCounter = 0;
                    }
                } else {
                    // Normal fluctuation
                    int change = random.nextInt(11) - 5; // -5 to +5
                    testHeartRate += change;

                    // Keep in reasonable range
                    if (testHeartRate < 60) testHeartRate = 60 + random.nextInt(10);
                    if (testHeartRate > 100) testHeartRate = 90 + random.nextInt(10);
                }

                currentHeartRate = testHeartRate;
                currentSteps += random.nextInt(5); // Simulate steps

                runOnUiThread(() -> {
                    updateUI();
                    sendDataToPhone();
                });

            } catch (InterruptedException e) {
                break;
            }
        }
    }).start();
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() == Sensor.TYPE_HEART_RATE) {
            currentHeartRate = (int) event.values[0];
            updateUI();
            sendDataToPhone();
        } else if (event.sensor.getType() == Sensor.TYPE_STEP_COUNTER) {
            currentSteps = (int) event.values[0];
            updateUI();
            sendDataToPhone();
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Not needed for this implementation
    }

    private void updateUI() {
        tvHeartRate.setText(currentHeartRate + " bpm");
        tvSteps.setText(String.valueOf(currentSteps));

        // Update status color based on heart rate
        if (currentHeartRate > 120 || currentHeartRate < 50) {
            tvStatus.setTextColor(getResources().getColor(android.R.color.holo_red_dark));
            if (useTestMode) {
                tvStatus.setText("Test: ALERT!");
            } else {
                tvStatus.setText("Status: ALERT!");
            }
        } else {
            tvStatus.setTextColor(getResources().getColor(android.R.color.holo_green_dark));
            if (useTestMode) {
                tvStatus.setText("Test Mode: Normal");
            } else {
                tvStatus.setText("Status: Normal");
            }
        }
    }

    private void sendDataToPhone() {
        PutDataMapRequest dataMap = PutDataMapRequest.create(DATA_PATH);
        dataMap.getDataMap().putInt("heartRate", currentHeartRate);
        dataMap.getDataMap().putInt("steps", currentSteps);
        dataMap.getDataMap().putLong("timestamp", System.currentTimeMillis());

        PutDataRequest request = dataMap.asPutDataRequest();
        request.setUrgent();

        Task<DataItem> putDataTask = dataClient.putDataItem(request);
        putDataTask.addOnSuccessListener(dataItem -> {
        // Data sent successfully
    });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (sensorManager != null && !useTestMode) {
            if (heartRateSensor != null) {
                sensorManager.registerListener(this, heartRateSensor, SensorManager.SENSOR_DELAY_NORMAL);
            }
            if (stepCounterSensor != null) {
                sensorManager.registerListener(this, stepCounterSensor, SensorManager.SENSOR_DELAY_NORMAL);
            }
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (sensorManager != null) {
            sensorManager.unregisterListener(this);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
        @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSION_REQUEST_CODE) {
            boolean allGranted = true;
            for (int result : grantResults) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    allGranted = false;
                    break;
                }
            }

            if (allGranted) {
                setupSensors();
            }
        }
    }
}

File: mobile/src/main/java/com/example/sobti/MainActivity.kt
package com.example.sobti

import android.Manifest
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.telephony.SmsManager
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.example.sobti.aws.AWSConfig
import com.example.sobti.aws.DynamoDBManager
import com.example.sobti.aws.DynamoDBManager.GetUserCallback
import com.example.sobti.aws.DynamoDBManager.UpdateCallback
import com.example.sobti.aws.DynamoDBManager.UserData
import com.google.android.gms.location.*
import com.google.android.gms.wearable.DataClient.OnDataChangedListener
import com.google.android.gms.wearable.DataEvent
import com.google.android.gms.wearable.DataEventBuffer
import com.google.android.gms.wearable.DataMapItem
import com.google.android.gms.wearable.Wearable

class MainActivity : AppCompatActivity(), OnDataChangedListener {

    private lateinit var tvHeartRate: TextView
    private lateinit var tvSteps: TextView
    private lateinit var tvLocation: TextView
    private lateinit var tvUserName: TextView
    private lateinit var tvStatus: TextView

    private lateinit var dbManager: DynamoDBManager
    private lateinit var prefs: SharedPreferences
    private var userEmail: String? = null
    private var emergencyNumber: String? = null

    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private var currentLocation: String = "Loading..."

    private var currentHeartRate = 0
    private var previousHeartRate = 0
    private var heartRateTrendCount = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        prefs = getSharedPreferences("SobtiPrefs", MODE_PRIVATE)
        userEmail = prefs.getString("user_email", "")

        // Initialize AWS
        AWSConfig.initialize(this)
        dbManager = DynamoDBManager(AWSConfig.getDDBClient())

        initViews()
        requestPermissions()
        setupLocationTracking()
        loadUserData()
        connectToWearable()
    }

    private fun initViews() {
        tvUserName = findViewById(R.id.tvUserName)
        tvHeartRate = findViewById(R.id.tvHeartRate)
        tvSteps = findViewById(R.id.tvSteps)
        tvLocation = findViewById(R.id.tvLocation)
        tvStatus = findViewById(R.id.tvStatus)
    }

    private fun requestPermissions() {
        val permissions = arrayOf(
            Manifest.permission.BODY_SENSORS,
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.SEND_SMS,
            Manifest.permission.ACTIVITY_RECOGNITION
        )
        ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_CODE)
    }

    private fun setupLocationTracking() {
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
            == PackageManager.PERMISSION_GRANTED
        ) {
            // ✅ Modern LocationRequest API
            val locationRequest = LocationRequest.Builder(
                Priority.PRIORITY_HIGH_ACCURACY,
                30000L // interval
            )
                .setMinUpdateIntervalMillis(15000L)
                .build()

            fusedLocationClient.requestLocationUpdates(
                locationRequest,
                object : LocationCallback() {
                    override fun onLocationResult(locationResult: LocationResult) {
                        val location = locationResult.lastLocation
                        location?.let {
                            currentLocation = String.format("%.6f, %.6f", it.latitude, it.longitude)
                            updateLocationUI()
                        }
                    }
                },
                null
            )
        }
    }

    private fun loadUserData() {
        dbManager.getUserData(userEmail, object : GetUserCallback {
            override fun onSuccess(userData: UserData) {
                runOnUiThread {
                    tvUserName.text = "Welcome, ${userData.name}!"
                    emergencyNumber = userData.emergencyNumber

                    if (userData.lastHeartRate > 0) {
                        tvHeartRate.text = "${userData.lastHeartRate} bpm"
                    }
                    if (userData.lastSteps > 0) {
                        tvSteps.text = userData.lastSteps.toString()
                    }
                    userData.lastLocation?.let {
                        currentLocation = it
                        updateLocationUI()
                    }
                }
            }

            override fun onError(e: Exception) {
                runOnUiThread {
                    Toast.makeText(
                        this@MainActivity,
                        "Error loading user data: ${e.message}",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
        })
    }

    private fun connectToWearable() {
        Wearable.getDataClient(this).addListener(this)
        tvStatus.text = "Status: Connected to Watch"
    }

    override fun onDataChanged(dataEvents: DataEventBuffer) {
        for (event in dataEvents) {
            if (event.type == DataEvent.TYPE_CHANGED) {
                val item = event.dataItem
                if (item.uri.path == WEAR_DATA_PATH) {
                    val dataMap = DataMapItem.fromDataItem(item).dataMap
                    val heartRate = dataMap.getInt("heartRate", 0)
                    val steps = dataMap.getInt("steps", 0)

                    runOnUiThread {
                        updateHealthData(heartRate, steps)
                    }
                }
            }
        }
    }

    private fun updateHealthData(heartRate: Int, steps: Int) {
        tvHeartRate.text = "$heartRate bpm"
        tvSteps.text = steps.toString()

        checkHeartRateTrend(heartRate)

        dbManager.updateHealthData(
            userEmail, heartRate, steps, currentLocation,
            object : UpdateCallback {
                override fun onSuccess() {}
                override fun onError(e: Exception?) {}
            })
    }

    private fun checkHeartRateTrend(heartRate: Int) {
        currentHeartRate = heartRate

        if (currentHeartRate > previousHeartRate && currentHeartRate > HIGH_HR_THRESHOLD) {
            heartRateTrendCount++
        } else if (currentHeartRate < previousHeartRate && currentHeartRate < LOW_HR_THRESHOLD) {
            heartRateTrendCount++
        } else {
            heartRateTrendCount = 0
        }

        if (heartRateTrendCount >= TREND_THRESHOLD) {
            triggerEmergency(heartRate)
            heartRateTrendCount = 0
        }

        previousHeartRate = currentHeartRate
    }

    private fun triggerEmergency(heartRate: Int) {
        if (emergencyNumber.isNullOrEmpty()) return

        val message = String.format(
            "SOBTI ALERT!\nAbnormal heart rate detected: %d bpm\nLocation: %s\nImmediate assistance needed!",
            heartRate, currentLocation
        )

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
            == PackageManager.PERMISSION_GRANTED
        ) {
            try {
                // ✅ Modern API to get SmsManager
                val smsManager = getSystemService(SmsManager::class.java)
                smsManager.sendTextMessage(emergencyNumber, null, message, null, null)

                runOnUiThread {
                    tvStatus.text = "Status: Emergency SMS Sent!"
                    tvStatus.setTextColor(getColor(android.R.color.holo_red_dark))
                    Toast.makeText(this, "Emergency alert sent!", Toast.LENGTH_LONG).show()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun updateLocationUI() {
        tvLocation.text = currentLocation
    }

    override fun onResume() {
        super.onResume()
        Wearable.getDataClient(this).addListener(this)
    }

    override fun onPause() {
        super.onPause()
        Wearable.getDataClient(this).removeListener(this)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == PERMISSION_REQUEST_CODE) {
            val allGranted = grantResults.all { it == PackageManager.PERMISSION_GRANTED }
            if (!allGranted) {
                Toast.makeText(
                    this,
                    "Some permissions denied. App may not work properly.",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    companion object {
        private const val PERMISSION_REQUEST_CODE = 100
        private const val WEAR_DATA_PATH = "/health_data"

        private const val TREND_THRESHOLD = 3
        private const val HIGH_HR_THRESHOLD = 120
        private const val LOW_HR_THRESHOLD = 50
    }
}
